stdlib/mexpr/type-lift.mc

```diff
@@ -5,19 +5,613 @@
+          match typeLiftType env t.ty with (env, ty) then
+            (env, TmMatch {{{{t with target = target}
+                                with thn = thn}
+                                with els = els}
+                                with ty = ty})
+          else never
+        else never
+      else never
+    else never
+end
+
+lang UtestTypeLift = TypeLift + UtestAst
+  sem typeLiftExpr (env : TypeLiftEnv) =
+  | TmUtest t ->
+    match typeLiftExpr env t.test with (env, test) then
+      match typeLiftExpr env t.expected with (env, expected) then
+        match typeLiftExpr env t.next with (env, next) then
+          match typeLiftType env t.ty with (env, ty) then
+            (env, TmUtest {{{{t with test = test}
+                                with expected = expected}
+                                with next = next}
+                                with ty = ty})
+          else never
+        else never
+      else never
+    else never
+end
+
+lang NeverTypeLift = TypeLift + NeverAst
+  sem typeLiftExpr (env : TypeLiftEnv) =
+  | TmNever t ->
+    match typeLiftType env t.ty with (env, ty) then
+      (env, TmNever {t with ty = ty})
+    else never
+end
 
 -----------
 -- TYPES --
 -----------
+
+lang UnknownTypeTypeLift = TypeLift + UnknownTypeAst
+  sem typeLiftType (env : TypeLiftEnv) =
+  | TyUnknown {} -> (env, TyUnknown {})
+end
+
+lang BoolTypeTypeLift = TypeLift + BoolTypeAst
+  sem typeLiftType (env : TypeLiftEnv) =
+  | TyBool {} -> (env, TyBool {})
+end
+
+lang IntTypeTypeLift = TypeLift + IntTypeAst
+  sem typeLiftType (env : TypeLiftEnv) =
+  | TyInt {} -> (env, TyInt {})
+end
+
+lang FloatTypeTypeLift = TypeLift + FloatTypeAst
+  sem typeLiftType (env : TypeLiftEnv) =
+  | TyFloat {} -> (env, TyFloat {})
+end
+
+lang CharTypeTypeLift = TypeLift + CharTypeAst
+  sem typeLiftType (env : TypeLiftEnv) =
+  | TyChar {} -> (env, TyChar {})
+end
+
+lang FunTypeTypeLift = TypeLift + FunTypeAst
+  sem typeLiftType (env : TypeLiftEnv) =
+  | TyArrow t ->
+    match typeLiftType env t.from with (env, from) then
+      match typeLiftType env t.to with (env, to) then
+        (env, TyArrow {from = from, to = to})
+      else never
+    else never
+end
+
+lang SeqTypeTypeLift = TypeLift + SeqTypeAst
+  sem typeLiftType (env : TypeLiftEnv) =
+  | TySeq t ->
+    match typeLiftType env t.ty with (env, ty) then
+      (env, TySeq {ty = ty})
+    else never
+end
+
+lang RecordTypeTypeLift = TypeLift + RecordTypeAst
+  sem typeLiftType (env : TypeLiftEnv) =
+  | TyRecord {fields = fields} & ty ->
+    if eqi (mapLength fields) 0 then
+      (env, ty)
+    else match mapLookup fields env.records with Some name then
+      (env, ntyvar_ name)
+    else
+      let recName = nameSym "Rec" in
+      let recTyVar = ntyvar_ recName in
+      let env = {{env with records = mapInsert fields recName env.records}
+                      with typeEnv = assocSeqInsert recName ty env.typeEnv} in
+      (env, recTyVar)
+end
```

Comment:
Shouldn't there be a recursive call to `typeLiftType` here somewhere? We need to lift out nested records before lifting the outer record.

Reply:
In `RecordTypeTypeLift`, that is.

Reply:
Ooops, yes that's right. This should be fixed now.